dts: flash_map: Automatically generate of the list of flash area

galak opened the pull request:
This Patch add functionality for automatic generation of the flash map
using DTS description. Automatic generation allows to replace
C-hardcoded flash_map.

We generate a set of defines based on the index of a partiion:
	#define DT_FLASH_AREA_<IDX>_OFFSET	0
	#define DT_FLASH_AREA_<IDX>_SIZE	131072
	#define DT_FLASH_AREA_<IDX>_DEV		"FLASH_CTRL"
	#define DT_FLASH_AREA_<IDX>_LABEL	MCUBOOT

Additionally we also define:
	#define DT_FLASH_AREA_NUM		4

and:
	#define DT_FLASH_AREA_<PARTNAME>_ID	0

Signed-off-by: Findlay Feng <i@fengch.me>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Commit:
dts: flash_map: Automatically generate of the list of flash area

This Patch add functionality for automatic generation of the flash map
using DTS description. Automatic generation allows to replace
C-hardcoded flash_map.

We generate a set of defines based on the index of a partiion:
	#define DT_FLASH_AREA_<IDX>_OFFSET	0
	#define DT_FLASH_AREA_<IDX>_SIZE	131072
	#define DT_FLASH_AREA_<IDX>_DEV		"FLASH_CTRL"
	#define DT_FLASH_AREA_<IDX>_LABEL	MCUBOOT

Additionally we also define:
	#define DT_FLASH_AREA_NUM		4

and:
	#define DT_FLASH_AREA_<PARTNAME>_ID	0

Signed-off-by: Findlay Feng <i@fengch.me>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

nvlsianpu committed:
tests/subsys/storage/flash_map: use auto-generated fa ID

After addition of flash_map auto-generation it is needed to
use generated flas_area name instead of hard-code.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@findlayfeng @nvlsianpu this is take the base 3 commits from @findlayfeng PR #8837 and reworks the basic dt extraction support.

I've built tested this but haven't run it on HW to make sure flash map is working properly.

nvlsianpu commented on include/flash_map.h:
```diff
@@ -42,6 +42,7 @@ struct flash_area {
 	u16_t pad16;
 	off_t fa_off;
 	size_t fa_size;
+	const char *fa_name;
```
I believe it fa_device_id is not needed anymore

nvlsianpu replied:
fa_name -> fa_dev_name

galak replied:
> I believe it fa_device_id is not needed anymore

Its not, was keeping it around until we decide on deprecation policy vs removal for out of tree flash_maps.

nvlsianpu reviewed (commented):
will test this

galak commented:
> @galak, would it not be better to use fa.dev_name instead of fa.name ?

Yeah, that makes more sense.

nvlsianpu commented:
With this patch flash area ID are not predefined - so need to use DT defined labels instead of explicite area numbers whenever call flash_area_open. FCB test needs corrections.

galak committed:
tests: boards: intel_s1000_crb: Use DT_FLASH_AREA_<FOO>_ID define

Rather than using hard coded 3, use DT_FLASH_AREA_IMAGE_SCRATCH_ID
that is now generated.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
tests: subsys: fs: fcb: Use DT_FLASH_AREA_<FOO>_ID define

Rather than using hard coded 2, use DT_FLASH_AREA_IMAGE_1_ID
that is now generated.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
> With this patch flash area ID are not predefined - so need to use DT defined labels instead of explicite area numbers whenever call flash_area_open. FCB test needs corrections.

Pushed changes to use DT_FLASH_AREA_* for in-tree users except `tests/subsys/settings/fcb/src/settings_test_fcb.c` & `subsys/settings/src/settings_init.c` as they use `CONFIG_SETTINGS_FCB_FLASH_AREA`.

nvlsianpu reviewed (approved):
looks good, works fine.

galak committed:
tests: subsys: settings: Use DT_FLASH_AREA_<FOO>_ID define

Set Kconfig symbol SETTINGS_FCB_FLASH_AREA based on the
DT_FLASH_AREA_<FOO>_ID define instead of being hard coded.

We replace 3 with DT_FLASH_AREA_IMAGE_SCRATCH_ID and
we replace 4 with DT_FLASH_AREA_STORAGE_ID.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
